libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1370,17 +1376,42 @@ void ExpressionCompiler::appendBitOperatorCode(Token::Value _operator)
 	}
 }
 
-void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator)
+void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator, Type const& _leftType, Type const& _rightType)
 {
-	BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Shift operators not yet implemented."));
+	// stack: rvalue lvalue
+
+	IntegerType const& leftType = dynamic_cast<IntegerType const&>(_leftType);
+	bool const c_leftSigned = leftType.isSigned();
+
+	// The RValue can be a RationalNumberType too.
+	bool c_rightSigned = false;
+	if (_rightType.category() == Type::Category::RationalNumber)
+	{
+		RationalNumberType const& rightType = dynamic_cast<RationalNumberType const&>(_rightType);
+		c_rightSigned = rightType.integerType()->isSigned();
+	}
+	else
+	{
+		IntegerType const& rightType = dynamic_cast<IntegerType const&>(_rightType);
+		c_rightSigned = rightType.isSigned();
+	}
+
+	// shift with negative rvalue throws exception
+	if (c_rightSigned) {
+		m_context << Instruction::DUP2 << (u256(1) << 255) << Instruction::AND << Instruction::ISZERO << Instruction::ISZERO;
```

Comment:
In case this is a rational number, the check should be done at compile time (inside the type checker). We could add an assert here.

Also perhaps shorter: `dup2 0 slt` (or `0 dup2 slt` - please cehck)